File: android/app/src/main/java/gg/yappy/app/ui/util/Time.kt
package gg.yappy.app.ui.util

import android.content.Context
import android.text.format.DateFormat
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale

/**
 * Whether to write clock times as 14:05 or 2:05 PM.
 *
 * This was hardcoded to 24-hour, so a phone set to 12-hour showed `00:51` on a
 * message sent at the moment its own status bar read `12:51`. Nothing else in
 * the app disagrees with the system like that, and it is the sort of detail
 * that makes software feel like it was built somewhere else.
 *
 * Held as a flag rather than resolved per call because these formatters are hit
 * once per visible bubble on every recomposition, and `is24HourFormat` goes to
 * a content provider. Refreshed when the activity resumes, which covers the
 * only realistic path: leaving to change it in system settings and coming back.
 */
object ClockStyle {
    @Volatile
    private var use24 = true

    fun refresh(context: Context) {
        use24 = DateFormat.is24HourFormat(context)
    }

    val time: DateTimeFormatter
        get() = if (use24) time24 else time12
}

private val time24: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm", Locale.getDefault())
private val time12: DateTimeFormatter = DateTimeFormatter.ofPattern("h:mm a", Locale.getDefault())
private val weekdayFormat: DateTimeFormatter = DateTimeFormatter.ofPattern("EEE", Locale.getDefault())
private val dateFormat: DateTimeFormatter = DateTimeFormatter.ofPattern("d MMM", Locale.getDefault())
private val fullDateFormat: DateTimeFormatter =
    DateTimeFormatter.ofPattern("d MMM yyyy", Locale.getDefault())

private fun parse(iso: String?): Instant? =
    iso?.let { runCatching { Instant.parse(it) }.getOrNull() }

/**
 * Conversation-list timestamps.
 *
 * Deliberately coarse and absolute past today: "3d ago" makes people do
 * arithmetic, while "Tue" and "14 Jan" are read at a glance. Only the last hour
 * gets relative treatment, where it genuinely is the more useful framing.
 */
fun relativeTime(iso: String?): String {
    val instant = parse(iso) ?: return ""
    val zone = ZoneId.systemDefault()
    val then = instant.atZone(zone)
    val now = Instant.now().atZone(zone)

    val minutes = ChronoUnit.MINUTES.between(then, now)
    return when {
        minutes < 1 -> "now"
        minutes < 60 -> "${minutes}m"
        then.toLocalDate() == now.toLocalDate() -> then.format(ClockStyle.time)
        then.toLocalDate() == now.toLocalDate().minusDays(1) -> "Yesterday"
        ChronoUnit.DAYS.between(then, now) < 7 -> then.format(weekdayFormat)
        then.year == now.year -> then.format(dateFormat)
        else -> then.format(fullDateFormat)
    }
}

/** Clock time on a message bubble. */
fun clockTime(iso: String?): String =
    parse(iso)?.atZone(ZoneId.systemDefault())?.format(ClockStyle.time).orEmpty()

// Built once like every other formatter in this file — `ofPattern` compiles
// the pattern, and this ran per day-separator per recomposition of the list.
private val SameYearDay = DateTimeFormatter.ofPattern("EEEE, d MMM")
private val OtherYearDay = DateTimeFormatter.ofPattern("d MMM yyyy")

/** Header for a day separator in the message list. */
fun dayLabel(iso: String?): String {
    val instant = parse(iso) ?: return ""
    val date = instant.atZone(ZoneId.systemDefault()).toLocalDate()
    val today = LocalDate.now()
    return when (date) {
        today -> "Today"
        today.minusDays(1) -> "Yesterday"
        else -> if (date.year == today.year) {
            date.format(SameYearDay)
        } else {
            date.format(OtherYearDay)
        }
    }
}

/** True when two messages fall on different calendar days in the local zone. */
fun crossesDay(previous: String?, current: String?): Boolean {
    val a = parse(previous)?.atZone(ZoneId.systemDefault())?.toLocalDate()
    val b = parse(current)?.atZone(ZoneId.systemDefault())?.toLocalDate() ?: return false
    return a != b
}

fun formatDuration(seconds: Int): String {
    val m = seconds / 60
    val s = seconds % 60
    return if (m >= 60) "%d:%02d:%02d".format(m / 60, m % 60, s) else "%d:%02d".format(m, s)
}

File: ios/Yappy/Components/Time.swift
import Foundation

/// Timestamp formatting.
///
/// Formatters are expensive to build and are reused rather than constructed per
/// row — a conversation list makes one call per visible cell on every state
/// change, and allocating a `DateFormatter` each time is measurable.
enum YappyTime {
    /// The API sends RFC 3339 with fractional seconds on some fields and without
    /// on others, and `ISO8601DateFormatter` matches only the format it was
    /// configured for. Both are tried rather than assuming.
    private static let isoWithFraction: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    private static let iso: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    private static let clock: DateFormatter = {
        let formatter = DateFormatter()
        // Locale-aware rather than a hard "HH:mm": on a device set to a 12-hour
        // clock, 24-hour times in the bubbles look like a bug.
        formatter.setLocalizedDateFormatFromTemplate("j:mm")
        return formatter
    }()

    private static let weekday: DateFormatter = {
        let formatter = DateFormatter()
        formatter.setLocalizedDateFormatFromTemplate("EEE")
        return formatter
    }()

    private static let shortDate: DateFormatter = {
        let formatter = DateFormatter()
        formatter.setLocalizedDateFormatFromTemplate("d MMM")
        return formatter
    }()

    private static let fullDate: DateFormatter = {
        let formatter = DateFormatter()
        formatter.setLocalizedDateFormatFromTemplate("d MMM yyyy")
        return formatter
    }()

    private static let dayHeader: DateFormatter = {
        let formatter = DateFormatter()
        formatter.setLocalizedDateFormatFromTemplate("EEEE d MMM")
        return formatter
    }()

    static func parse(_ iso8601: String?) -> Date? {
        guard let iso8601, !iso8601.isEmpty else { return nil }
        return isoWithFraction.date(from: iso8601) ?? iso.date(from: iso8601)
    }

    static func now() -> String { isoWithFraction.string(from: Date()) }

    /// A specific instant, for a request body. `now()` is this with `Date()`.
    static func string(from date: Date) -> String { isoWithFraction.string(from: date) }

    /// Conversation-list timestamps.
    ///
    /// Deliberately coarse and absolute past today — "3d ago" makes people do
    /// arithmetic, while "Tue" and "14 Jan" are read at a glance. Only the last
    /// hour gets relative treatment, where it genuinely is the more useful
    /// framing.
    static func relative(_ iso8601: String?) -> String {
        guard let date = parse(iso8601) else { return "" }
        let calendar = Calendar.current
        let now = Date()

        let minutes = Int(now.timeIntervalSince(date) / 60)
        if minutes < 1 { return "now" }
        if minutes < 60 { return "\(minutes)m" }
        if calendar.isDateInToday(date) { return clock.string(from: date) }
        if calendar.isDateInYesterday(date) { return "Yesterday" }

        let days = calendar.dateComponents([.day], from: date, to: now).day ?? 0
        if days < 7 { return weekday.string(from: date) }
        if calendar.component(.year, from: date) == calendar.component(.year, from: now) {
            return shortDate.string(from: date)
        }
        return fullDate.string(from: date)
    }

    /// Clock time on a message bubble.
    static func clockTime(_ iso8601: String?) -> String {
        guard let date = parse(iso8601) else { return "" }
        return clock.string(from: date)
    }

    /// Header for a day separator in the message list.
    static func dayLabel(_ iso8601: String?) -> String {
        guard let date = parse(iso8601) else { return "" }
        let calendar = Calendar.current
        if calendar.isDateInToday(date) { return "Today" }
        if calendar.isDateInYesterday(date) { return "Yesterday" }
        if calendar.component(.year, from: date) == calendar.component(.year, from: Date()) {
            return dayHeader.string(from: date)
        }
        return fullDate.string(from: date)
    }

    /// True when two messages fall on different calendar days in the local zone.
    static func crossesDay(_ previous: String?, _ current: String?) -> Bool {
        guard let currentDate = parse(current) else { return false }
        guard let previousDate = parse(previous) else { return true }
        return !Calendar.current.isDate(previousDate, inSameDayAs: currentDate)
    }

    static func duration(_ seconds: Int) -> String {
        let minutes = seconds / 60
        let secs = seconds % 60
        if minutes >= 60 {
            return String(format: "%d:%02d:%02d", minutes / 60, minutes % 60, secs)
        }
        return String(format: "%d:%02d", minutes, secs)
    }
}
